File: JavaScript/CheckoutSystem.js
const prompt = require("prompt-sync")();
const now = new Date();

let customers = [];
let cashiers = [];


function getCurrentDateTime(){
  return new Date().toLocaleString();
}

function addCustomer(customerName, customerPhone){
    return customers.push({customerName, customerPhone});
}
/*
function addToCart(item, qty, pricePerUnit){
    return cart.push({item, qty, pricePerUnit});
}
*/
function addCashier(cashierName){
    return cashiers.push(cashierName);
}

function getItemPrice(qty, pricePerUnit){
    return qty * pricePerUnit;
}

function getCartTotal(cart){
    let cartTotal = 0;
    for(let item of cart){
        cartTotal += item.qty * item.pricePerUnit;
    }
    return cartTotal;
}

function giveDiscount(cartTotal, discountRate){
    let discount = cartTotal * (discountRate / 100);
    let amountAfterDiscount = cartTotal - discount;
    return {discount,
            amountAfterDiscount};
}

function takeVAT(amountAfterDiscount, vatRate){
    let vat = amountAfterDiscount * (vatRate / 100);
    let amountAfterVAT = amountAfterDiscount + vat
    return {vat, amountAfterVAT};
}

function getNetAmount(cartTotal, discount, vat){
    let netAmount = (cartTotal - discount) + vat;
    return netAmount
}

function getInvoice(customerName, customerPhone, cashierName, cart, cartTotal, discountRate, discount, vatRate, vat, netAmount){
    let invoice = `
____________________INVOICE_______________________________
SEMICOLON STORES
MAIN BRANCH
LOCATION: 312, HERBERT MACAULAY WAY, SABO YABA, LAGOS.
TEL: 08089765432
Date: ${(getCurrentDateTime())}
Cashier: ${cashierName}
Customer: ${customerName}   Ph.No: ${customerPhone}
==========================================================
ITEM(s)         QTY         PRICE               TOTAL(NGN)
`;
    for(let item of cart){
        let itemTotal = item.qty * item.pricePerUnit;
`----------------------------------------------------------`
        invoice += `\n${item.item.padEnd(18)} ${String(item.qty).padEnd(9)} ${String(item.pricePerUnit).padEnd(10)} ${itemTotal}`
    }
    invoice += `
----------------------------------------------------------
                            Sub Total:      ${cartTotal}
                            Discount:       ${discount}
                        VAT @ ${vatRate}%:  ${vat}
==========================================================
                            Bill Total:     ${netAmount}
==========================================================
        THIS IS NOT A RECEIPT. KINDLY PAY ${netAmount}
==========================================================
`;
    console.log(invoice);
}

function getBalance(customerPay, netAmount){
    let balance = customerPay - netAmount;
    return balance;
}

function getReceipt(customerName, customerPhone, cashierName, cart, cartTotal, discountRate, discount, vatRate, vat, netAmount, customerPay, balance){
    let receipt = `
________________________RECEIPT___________________________
SEMICOLON STORES
MAIN BRANCH
LOCATION: 312, HERBERT MACAULAY WAY, SABO YABA, LAGOS.
TEL: 08089765432
Date: ${(getCurrentDateTime())}
Cashier: ${cashierName}
Customer: ${customerName}   Ph.No: ${customerPhone}
==========================================================
ITEM(s)         QTY         PRICE               TOTAL(NGN)
`;
    for(let item of cart){
        let itemTotal = item.qty * item.pricePerUnit;
`----------------------------------------------------------`
        receipt += `\n${item.item.padEnd(18)} ${String(item.qty).padEnd(9)} ${String(item.pricePerUnit).padEnd(10)} ${itemTotal}`
    }
    receipt += `
----------------------------------------------------------
                            Sub Total:      ${cartTotal}
                            Discount:       ${discount}
                        VAT @ ${vatRate}%:  ${vat}
==========================================================
                            Bill Total:     ${netAmount}
                            Amount Paid:    ${customerPay}
                            Balance:        ${balance}
==========================================================
                THANK YOU FOR YOUR PATRONAGE
==========================================================
`;
    console.log(receipt);
}

//...
while(true){
    let cart = [];
    let customerName = prompt("Enter customer's name: ");
    let customerPhone = prompt("Enter customer's phoneNo.: ");
    addCustomer(customerName, customerPhone);

let nav = "yes";
    while(true){
        let item = prompt(`What did customer ${customerName} buy?: `);
        let qty = parseInt(prompt(`How many quantities of ${item} did customer ${customerName} buy: `));
        let pricePerUnit = parseFloat(prompt(`What is the unit price of ${item}: `));
        cart.push({item, qty, pricePerUnit});
        nav = prompt("Add more items?(yes or no): ").toLowerCase();
        if(nav === "no"){
            break;
        }
    }

    let cartTotal = getCartTotal(cart);

    let cashierName = prompt("You are the cashier, enter your name: ");
    addCashier(cashierName);

    let discountRate = parseFloat(prompt("Enter discount rate (if none, enter 0): "))
    let {discount, amountAfterDiscount} = giveDiscount(cartTotal, discountRate);
    let vatRate = parseFloat(prompt("Enter VAT rate: "));
    let {vat, amountAfterVAT} = takeVAT(amountAfterDiscount, vatRate);
    let netAmount = getNetAmount(cartTotal, discount, vat);

    getInvoice(customerName, customerPhone, cashierName, cart, cartTotal, discountRate, discount, vatRate, vat, netAmount);

    let customerPay = parseFloat(prompt(`How much did customer ${customerName} give to you?: `));
    let balance = getBalance(customerPay, netAmount);

    getReceipt(customerName, customerPhone, cashierName, cart, cartTotal, discountRate, discount, vatRate, vat, netAmount, customerPay, balance);

//    console.log(`Customer Paid: ${customerPay}`);
//    console.log(`Balance: ${balance}`);

    let anotherCustomer = prompt("Serve another customer? (yes/no): ").toLowerCase();
    if(anotherCustomer !== "yes"){
        console.log("Closing register... Goodbye!");
        break;
    }
}

File: Python/checkout_system.py
from datetime import datetime
from datetime import timedelta

customers = []
cashiers = []

def get_current_datetime():
	return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

def add_customer(customer_name, customer_phone):
    customers.append({"name": customer_name, "phone": customer_phone})
    return len(customers)

def add_cashier(cashier_name):
    cashiers.append(cashier_name)
    return len(cashiers)

def get_item_price(qty, price_per_unit):
    return qty * price_per_unit

def get_cart_total(cart):
    cart_total = 0
    for item in cart:
        cart_total += item["qty"] * item["price_per_unit"]
    return cart_total

def give_discount(cart_total, discount_rate):
    discount = cart_total * (discount_rate / 100)
    amount_after_discount = cart_total - discount
    return discount, amount_after_discount

def take_vat(amount_after_discount, vat_rate):
    vat = amount_after_discount * (vat_rate / 100)
    amount_after_vat = amount_after_discount + vat
    return vat, amount_after_vat

def get_net_amount(cart_total, discount, vat):
    net_amount = (cart_total - discount) + vat
    return net_amount

def get_invoice(customer_name, customer_phone, cashier_name, cart, cart_total, discount_rate, discount, vat_rate, vat, net_amount):
    print("_______________________INVOICE______________________________")
    print("SEMICOLON STORES\nMAIN BRANCH\nLOCATION: 312, HERBERT MACAULAY WAY, SABO YABA, LAGOS.\nTEL: 08089765432")
    print("Date: ", get_current_datetime())
    print("Cashier: ", cashier_name)
    print("Customer: ", customer_name)
    print("Ph.No: ", customer_phone)
    print("============================================================")
    print("ITEM(s)\tQTY\tPRICE\tTOTAL(NGN)")
    for item in cart:
        item_total = item["qty"] * item["price_per_unit"]
        print(f"{item["item"]:>16} {str(item["qty"]):>10} {str(item["price_per_unit"]):>10} {item_total}")
    print("------------------------------------------------------------")
    print(f"{'SubTotal: ':>40}", cart_total)
    print(f"{'Discount: ':>40}", discount)
    print(f"{'VAT: ':>40}", vat)
    print("============================================================")
    print(f"{'Bill Total: ':>40}", net_amount)
    print("============================================================")
    print(f"\t{'THIS IS NOT A RECEIPT. KINDLY PAY'}", net_amount)
    print("============================================================")

def get_balance(customer_pay, net_amount):
    balance = customer_pay - net_amount
    return balance

def get_receipt(customer_name, customer_phone, cashier_name, cart, cart_total, discount_rate, discount, vat_rate, vat, net_amount, customer_pay, balance):
    print("_______________________INVOICE______________________________")
    print("SEMICOLON STORES\nMAIN BRANCH\nLOCATION: 312, HERBERT MACAULAY WAY, SABO YABA, LAGOS.\nTEL: 08089765432")
    print("Date: ", get_current_datetime())
    print("Cashier: ", cashier_name)
    print("Customer: ", customer_name)
    print("Ph.No: ", customer_phone)
    print("============================================================")
    print("ITEM(s)\tQTY\tPRICE\tTOTAL(NGN)")
    for item in cart:
        item_total = item["qty"] * item["price_per_unit"]
        print(f"{item["item"]:>16} {str(item["qty"]):>10} {str(item["price_per_unit"]):>10} {item_total}")
    print("------------------------------------------------------------")
    print(f"{'SubTotal: ':>40}", cart_total)
    print(f"{'Discount: ':>40}", discount)
    print(f"{'VAT: ':>40}", vat)
    print("============================================================")
    print(f"{'Bill Total: ':>40}", net_amount)
    print(f"{'Amount Paid: ':>40}", customer_pay)
    print(f"{'Balance: ':>40}", balance)
    print("============================================================")
    print("\tTHANK YOU FOR YOUR PATRONAGE")
    print("============================================================")


#...

while True:
    cart = []
    customer_name = input("Enter customer's name: ")
    customer_phone = input("Enter customer's phoneNo.: ")
    add_customer(customer_name, customer_phone)

    nav = "yes"
    while True:
        item = input(f"What did {customer_name} buy?: ")
        qty = int(input(f"How many quantities of {item} did customer {customer_name} buy?: "))
        price_per_unit = float(input(f"What is the unit price of {item}?: "))
        cart.append({"item": item, "qty": qty, "price_per_unit": price_per_unit})
        nav = input("Add more items? (yes or no): ").lower()
        if nav == "no":
            break

    cart_total = get_cart_total(cart)
    cashier_name = input("You are the cashier, enter your name: ")
    add_cashier(cashier_name)

    discount_rate = float(input("Enter discount rate (if none, enter 0): "))
    discount, amount_after_discount = give_discount(cart_total, discount_rate)
    #amount_after_discount = give_discount(cart_total, discount_rate)
    vat_rate = float(input("Enter VAT rate: "))
    vat, amount_after_vat = take_vat(amount_after_discount, vat_rate);
    #amount_after_vat = take_vat(amount_after_discount, vat_rate);
    net_amount = get_net_amount(cart_total, discount, vat);

    get_invoice(customer_name, customer_phone, cashier_name, cart, cart_total, discount_rate, discount, vat_rate, vat, net_amount)

    customer_pay = int(input("How much did the customer give you?: "))
    balance = get_balance(customer_pay, net_amount)

    get_receipt(customer_name, customer_phone, cashier_name, cart, cart_total, discount_rate, discount, vat_rate, vat, net_amount, customer_pay, balance)

    another_customer = input("Serve another customer? (yes/no): ").lower()
    if another_customer == "no":
        print("Closing register...Goodbye!")
        break
